Readme, Packaging, project structure, gh actions, fixes and others

aahnik opened the pull request:
A lot of planned work and related work is incomplete in this PR. I will do those in upcoming PRs.

### Make the readme accurate more useful to all stakeholders
- [x] update readme
- [x] move developer's guide from readme to `.github/CONTRIBUTING.md`. Full guide to clone/setup pre-commit/local database server/setup vs code/links to gh docs
  

### Proper Packaging and imports
- [x] create a python package `bhagavad-gita-api`
- [x] fix missing imports, make all imports absolute (currently it's mixed)

### Separation of concerns

- [x] keep data in a separate gh repo 
- [x] when the API server is started automatically download and insert data in the database if the deployer's database is empty


### Give convenience to self hosters

- [x] give deployer full freedom to use any SQL database. If no SQLAlchemy URI is provided, use a local SQLite by default
- [x] create Dockerfile


### Enforce code quality using pre-commit and CI
- [x] use pre-commit on ci to enforce code quality (using gh actions)
- [x] allow python versions above 3.8 (less than python 4 as 4 would have breaking changes)

ETA: ~June 29th~, ~6 PM IST~, ~9 PM IST~ | June 30th, 6 PM IST

aahnik committed:
update project structure

-> move data to a seperate repo
-> update the way modules are organized
-> move some code around
-> create a package bhagavad_gita_api
-> fix code quality errors

name of repo = name in PyPI = name of python package = name in DockerHub

aahnik committed:
move gita.png to .github folder

aahnik committed:
update pyproject.toml

-> move isort configuration here
-> make black and isort play well together
-> add package metadata
-> move many deps to dev deps

aahnik committed:
delete .python-version and run.py

aahnik committed:
add markdownlint to pre-commit config

aahnik committed:
use the standard ignore files

aahnik committed:
create Dockerfile

aahnik committed:
create Makefile for common cli utils

aahnik committed:
create a github action workflow for code quality

aahnik committed:
add a detailed friendly guide for contributors

aahnik committed:
update readme

aahnik committed:
commit poetry lock file to source control

samanyougarg commented on README.md:
```diff
@@ -9,88 +10,105 @@
+
+Here is a list of interesting projects using this API.
+
+- [bhagavadGita.io](https://bhagavadgita.io)
```
Minor typo. Can we change it from `bhagavadGita.io` to `BhagavadGita.io`?

samanyougarg commented on README.md:
```diff
@@ -9,88 +10,105 @@
+| Name                      | Description                           | Default     |
+| ------------------------- | ------------------------------------- | ----------- |
+| `TESTER_API_KEY`          | The API key for testing.              | `None`      |
+| `SQLALCHEMY_DATABASE_URI` | The DSN for your database connection. | `sqlite://` (in memory SQLite db)|
```
Currently, only Postgres DSN will work as per the configuration in `config.py`. We will need to update the type for `SQLALCHEMY_DATABASE_URI` to support other databases.

samanyougarg commented on .github/CONTRIBUTING.md:
````diff
@@ -0,0 +1,65 @@
+    ```shell
+    pre-commit install
+    ```
+
````
Can we add a step for setting up the environment variables before running the app? Otherwise, the user will get an error.

samanyougarg commented on .github/CONTRIBUTING.md:
````diff
@@ -0,0 +1,65 @@
+    pre-commit install
+    ```
+
+6. To start the server with hot reload,
````
Similarly, the instructions for initializing the database are missing here.

samanyougarg reviewed (commented):
Thank you very much for taking on this task. Everything looks great 👍 

A couple of minor things:
1. There are some comments regarding the README/CONTRIBUTING files. Please fix those.
2. Can we move the Pydantic schemas in a separate folder called `schemas` instead of having it along with the SQLAlchemy models in the `models` folder.
3. So, I believe this package will let the user set up the application in their own local environment or deploy it on their own servers. 
Do you think we should have another package that is basically a wrapper around our API and lets the user access the API via a Python library just like https://pypi.org/project/pygita/?

aahnik committed:
make multiple requested changes and others

    -> update contributing guide. Tell readers about setting env vars
    -> create data on startup, if missing
    -> create custom Dsn class inspired by PostgresDsn
    the new custom class SqlDsn will allow all types of sql db dsn.
    -> fix capitalization of spelling in readme

    -> change the linux permission for .gitignore, LICENSE and README.md
    its funny that someone made them executable
    i just did
    chmod -x filename
